### PR TITLE
WIP: ByteVec string implementation

### PR DESCRIPTION
**Pull request by StefanKarpinski**
@JeffBezanson, @Keno, @vtjnash, I'm having trouble debugging what's going on here – I'm pushing it in an incomplete state with lots of debugging output in hopes someone can help figure out what's going on here. So this croaks during system image build as soon as it hits code that uses a Regex:

``` jl
libc.jl
"A"
ASCIIString
Int128(0xffffffffffffffeb0000000106c75770)
Int128(0x7c6a7c417c612825295e7c5d255e5b28)
()
"B"
ASCIIString
Int128(0x7c6a7c417c612825295e7c5d255e5b28)
"C"
ASCIIString
Int128(0x7c6a7c417c612825295e7c5d255e5b28)
"D"
Int128(0x7c6a7c417c612825295e7c5d255e5b28)

signal (11): Segmentation fault: 11
length at bytevec.jl:18
print_to_string at string.jl:24
compile at pcre.jl:95
```

What you can see here is that it's hitting the `r"..."` macro just fine here:

https://github.com/JuliaLang/julia/blob/251202fa8f8445053f66fd/base/regex.jl#L72-L78

At this point, `pattern` is an `ASCIIString` with a data field of `0xffffffffffffffeb0000000106c75770`, as expected – it's a negative length and a pointer to the actual string data, which starts with `0x7c6a7c417c612825295e7c5d255e5b28` – or in other words, the data `([^%]|^)%(a|A|j|`, which is the initial part of a regex pattern that gets passed to the `r_str` macro. So far so good. This in turn invokes the `Regex` constructor, here:

https://github.com/JuliaLang/julia/blob/251202fa8f8445053f66fd/base/regex.jl#L18-L32

But now, surprisingly, `pattern` is no longer `0xffffffffffffffeb0000000106c75770` but seems to be the value that this used to point to – i.e. `0x7c6a7c417c612825295e7c5d255e5b28`. I have no idea how this is happening, since the same value of `pattern` is passed from the macro to the constructor function. Any ideas?

### CODE REVIEW

File: src/julia.h [hunk 1]
```diff
@@ -615,8 +630,7 @@ static inline void gc_wb_back(void *ptr) // ptr isa jl_value_t*
 #define jl_gc_unpreserve()
 #define jl_gc_n_preserved_values() (0)
 
-#define allocb(nb)    malloc(nb)
-DLLEXPORT jl_value_t *allocobj(size_t sz);
+#define allocb(nb) malloc(nb)
```
Comment (vtjnash) [inline src/julia.h]: why is this line changing?

File: src/julia.h [hunk 2]
```diff
@@ -179,6 +179,20 @@ STATIC_INLINE int jl_array_ndimwords(uint32_t ndims)
     return (ndims < 3 ? 0 : ndims-2);
 }
 
+typedef struct {
+    union {
```
Comment (vtjnash) [inline src/julia.h]: add `JL_DATA_TYPE` here to hint to the reader that this is a subtype of `jl_value_t`

File: src/intrinsics.cpp [hunk 3]
```diff
@@ -918,6 +920,180 @@ static Value *emit_intrinsic(intrinsic f, jl_value_t **args, size_t nargs,
     Value *den;
     Value *typemin;
     switch (f) {
+    HANDLE(bytevec_ref,2) {
```
Comment (vtjnash) [inline src/intrinsics.cpp]: these should probably be builtin function rather than intrinsics, unless jeff has some other preference?

File: src/intrinsics.cpp [hunk 4]
```diff
@@ -918,6 +920,180 @@ static Value *emit_intrinsic(intrinsic f, jl_value_t **args, size_t nargs,
...
+        BasicBlock *there = BasicBlock::Create(getGlobalContext(), "there", ctx->f);
+        BasicBlock *oob   = BasicBlock::Create(getGlobalContext(), "oob",   ctx->f);
+        BasicBlock *cont  = BasicBlock::Create(getGlobalContext(), "cont",  ctx->f);
+
+        // branch: "here" or "there"
+        Value *is_here = builder.CreateICmpSGE(b, ConstantInt::get(b->getType(), 0));
+        Value *here_len = builder.CreateZExt(
+            builder.CreateTrunc(
+                builder.CreateLShr(
+                    b, ConstantInt::get(b->getType(), 8*(2*sizeof(void*)-1))
```
Comment (vtjnash) [inline src/intrinsics.cpp]: please rewrite this expression using `offsetof` and `sizeof` + comments so the reader can figure out what that calculation does

File: src/intrinsics.cpp [hunk 5]
```diff
@@ -918,6 +920,180 @@ static Value *emit_intrinsic(intrinsic f, jl_value_t **args, size_t nargs,
...
+                builder.CreateAnd(builder.CreateNot(is_here), builder.CreateICmpULT(i, there_len)),
+                there, oob
+            );
+
+            // raise out-of-bounds error
+            builder.SetInsertPoint(oob);
+            builder.CreateCall2(
+                prepare_call(jlthrow_line_func),
+                tbaa_decorate(tbaa_const, builder.CreateLoad(prepare_global(jlemptysvec_var))),
+                ConstantInt::get(T_int32, ctx->lineno)
```
Comment (vtjnash) [inline src/intrinsics.cpp]: i don't think `throw(Base.svec())` is a particularly legal statement in Julia (although it does seem to work at the repl). there are a lot of bounds errors functions that you can call however (see `emit_bounds_check` in cgutils.cpp)

File: src/julia.h (hunk 2 above)
Comment (StefanKarpinski) [inline src/julia.h]: It's not – it's just a struct that's mirrored by the `ByteVec` immutable on the Julia side.

File: src/intrinsics.cpp (hunk 3 above)
Comment (StefanKarpinski) [inline src/intrinsics.cpp]: We don't want the overhead of doing a C call to pull bytes out of a buffer and we also need to be able to turn off bounds checking, which, AFAIK, there's no way to do for builtins. I'm pretty sure that Jeff wants to get rid of builtins, not add more of them. Really, the *only* reason these can't be written in pure Julia is because of the bounds checking – we can't turn it off for pure Julia code.

File: src/array.c [hunk 6]
```diff
@@ -344,14 +344,26 @@ jl_array_t *jl_pchar_to_array(const char *str, size_t len)
 jl_value_t *jl_array_to_string(jl_array_t *a)
 {
     // TODO: check type of array?
-    jl_datatype_t *string_type = u8_isvalid((char*)a->data, jl_array_len(a)) == 1 ? // ASCII
+    uint8_t *data = (uint8_t*)a->data;
+    size_t n = jl_array_len(a);
+    jl_datatype_t *string_type = u8_isvalid((char*)data,n) == 1 ?
         jl_ascii_string_type : jl_utf8_string_type;
-    jl_value_t *s = (jl_value_t*)alloc_1w();
+    jl_value_t *s = (jl_value_t*)alloc_2w();
```
Comment (vtjnash) [inline src/array.c]: alloc_4w on x32

File: src/julia.h (hunk 1 above)
Comment (StefanKarpinski) [inline src/julia.h]: unrelated whitespace cleanup

File: src/intrinsics.cpp (hunk 5 above)
Comment (StefanKarpinski) [inline src/intrinsics.cpp]: Yes, it's completely wrong but it works (and gives a useless error message). I needed a placeholder here when rebasing after `jlboundserr_var` disappeared and haven't spent the time to figure out what the new way to raise this error is. This is part of the wip-ness of the pull request. Thanks for the thorough review though.

File: src/array.c [hunk 7]
```diff
@@ -344,14 +344,26 @@ jl_array_t *jl_pchar_to_array(const char *str, size_t len)
...
 }
 
+static jl_bytevec_struct_t jl_shared_bytevec[1024];
+
+const char *jl_bytestring_ptr(jl_value_t *s) {
+    jl_bytevec_struct_t *b = (jl_bytevec_struct_t*)s;
+    if (b->there.neglen < 0)
+        return b->there.data;
+    jl_shared_bytevec[0] = *b;
+    return jl_shared_bytevec[0].here.data;
```
Comment (vtjnash) [inline src/array.c]: why is there a random no-op pointer copy here?

File: src/array.c [hunk 8]
```diff
@@ -344,14 +344,26 @@ jl_array_t *jl_pchar_to_array(const char *str, size_t len)
...
         jl_ascii_string_type : jl_utf8_string_type;
-    jl_value_t *s = (jl_value_t*)alloc_1w();
+    jl_value_t *s = (jl_value_t*)alloc_2w();
+    *(jl_bytevec_struct_t*)s = jl_bytevec(data,n);
     jl_set_typeof(s, string_type);
-    jl_set_nth_field(s, 0, (jl_value_t*)a);
     return s;
 }
 
+static jl_bytevec_struct_t jl_shared_bytevec[1024];
```
Comment (vtjnash) [inline src/array.c]: ???

File: src/julia.h [hunk 9]
```diff
@@ -179,6 +179,20 @@ STATIC_INLINE int jl_array_ndimwords(uint32_t ndims)
...
+        struct {
+            char data[2*sizeof(void*)-1];
+            uint8_t length;
+        } here;
+        struct {
+            char *data;
+            intptr_t neglen;
+        } there;
+        __uint128_t value;
```
Comment (StefanKarpinski) [inline src/julia.h]: @vtjnash, this was just temporarily inserted to see if it was a calling convention issue that could be fixed like this; normally this structure is two words on all platforms.

File: src/array.c (hunk 6 above)
Comment (StefanKarpinski) [inline src/array.c]: See comment below – this struct is normally two words on all platforms.

File: src/alloc.c [hunk 10]
```diff
@@ -308,6 +308,22 @@ DLLEXPORT jl_value_t *jl_new_struct_uninit(jl_datatype_t *type)
     return jv;
 }
 
+DLLEXPORT jl_bytevec_struct_t jl_bytevec(const uint8_t *data, size_t n)
+{
+    jl_bytevec_struct_t b;
+    if (n < 2*sizeof(void*)) {
```
Comment (vtjnash) [inline src/alloc.c]: `n  < sizeof(b.here.data)` to leave space for the null byte?

File: src/alloc.c [hunk 11]
```diff
@@ -308,6 +308,22 @@ DLLEXPORT jl_value_t *jl_new_struct_uninit(jl_datatype_t *type)
...
+        memcpy(b.here.data, data, n);
+        memset(b.here.data + n, 0, (2*sizeof(void*)-1) - n);
```
Comment (vtjnash) [inline src/alloc.c]: `sizeof(b.here.data) - n`

File: src/array.c (hunk 8 above)
Comment (StefanKarpinski) [inline src/array.c]: Just trying to rule out someone getting a pointer to this and accidentally smashing through the memory after it. Still crashes, so that's not the issue.

File: src/alloc.c [hunk 12]
```diff
@@ -308,6 +308,22 @@ DLLEXPORT jl_value_t *jl_new_struct_uninit(jl_datatype_t *type)
     return jv;
 }
 
+DLLEXPORT jl_bytevec_struct_t jl_bytevec(const uint8_t *data, size_t n)
```
Comment (vtjnash) [inline src/alloc.c]: it's weird to be returning a jl_value_t by value

File: src/array.c [hunk 13]
```diff
@@ -344,14 +344,26 @@ jl_array_t *jl_pchar_to_array(const char *str, size_t len)
...
 {
     // TODO: check type of array?
-    jl_datatype_t *string_type = u8_isvalid((char*)a->data, jl_array_len(a)) == 1 ? // ASCII
+    uint8_t *data = (uint8_t*)a->data;
+    size_t n = jl_array_len(a);
+    jl_datatype_t *string_type = u8_isvalid((char*)data,n) == 1 ?
         jl_ascii_string_type : jl_utf8_string_type;
-    jl_value_t *s = (jl_value_t*)alloc_1w();
+    jl_value_t *s = (jl_value_t*)alloc_2w();
+    *(jl_bytevec_struct_t*)s = jl_bytevec(data,n);
```
Comment (vtjnash) [inline src/array.c]: jl_bytevec allocates memory, so you need to zero the `s` struct, root it, and set it's type before this point

File: src/array.c (hunk 7 above)
Comment (StefanKarpinski) [inline src/array.c]: It's writing the value of `b` to the `jl_shared_bytevec` global and then returning a pointer to the data portion of that. At least that's the intention.

File: src/array.c (hunk 13 above)
Comment (vtjnash) [inline src/array.c]: you would also need a call to `gc_wb(s, s->field0)` here

(or reverse the order of the alloc and bytevec calls, but then the bytevec isn't rooted during the alloc call)

File: src/alloc.c (hunk 10 above)
Comment (StefanKarpinski) [inline src/alloc.c]: No, the length of the string is encoded in the last byte.

File: src/alloc.c (hunk 11 above)
Comment (StefanKarpinski) [inline src/alloc.c]: Yeah, that's clearer. Should be the exact same thing though.

File: src/intrinsics.cpp (hunk 3 above)
Comment (JeffBezanson) [inline src/intrinsics.cpp]: I'm fine with an intrinsic for now. I really don't want to add more "builtins", and it would be best to gradually remove them if possible.

@vtjnash excellent code review, thanks.

File: src/julia.h (hunk 1 above)
Comment (vtjnash) [inline src/julia.h]: the `DLLEXPORT allocobj` line also disappeared

File: src/intrinsics.cpp (hunk 3 above)
Comment (StefanKarpinski) [inline src/intrinsics.cpp]: In case it's not obvious, this code is not meant to be in production-quality state – this is a request for debugging help.

File: src/julia.h (hunk 1 above)
Comment (StefanKarpinski) [inline src/julia.h]: It was declared twice:

https://github.com/JuliaLang/julia/blob/b458ea3937f6bfa7b84a1eb6769f487271664f60/src/julia.h#L563
https://github.com/JuliaLang/julia/blob/b458ea3937f6bfa7b84a1eb6769f487271664f60/src/julia.h#L619

Since it only needs to be declared once, I deleted one of the declarations.

File: src/julia.h (hunk 2 above)
Comment (vtjnash) [inline src/julia.h]: immutable objects are not, in general, visible to the garbage collector

Comment (StefanKarpinski) [inline src/julia.h]: This type only exists to make it more convenient to write C functions that interact with the *data* part of the `ByteVec` type. Garbage collection is handled here:

https://github.com/JuliaLang/julia/blob/ed760e40e972fc1c6be/src/gc.c#L1631-L1636

GC needs to know whether or not to follow the pointer, which depends on the value.

Comment (vtjnash) [inline src/julia.h]: the GC won't see the pointer (or the bytevec), in general, if it is declared immutable. putting it inside another object (such as an ASCIIString) is sufficient to hide it from GC tracking.

Comment (StefanKarpinski) [inline src/julia.h]: The type is marked as not being pointer-free, despite being immutable,
which seems to work fine. In any case, I don't think this is likely to be
what's causing the problem here.

**Comment by vtjnash** [comment]
there seems to be some confusion in the codegen about the proper layout for this object composed of a recursive immutable:
```
# pattern.data.x
  %27 = extractvalue %ASCIIString %1, 0, !dbg !20
  %28 = bitcast %jl_value_t* %27 to i128*, !dbg !20
  %29 = load i128* %28, align 8, !dbg !20, !tbaa !21, !julia_type !23
```
```
(lldb) p *(jl_datatype_t*)jl_typeof(args[1])
(jl_datatype_t) $5 = {
  fieldptr0 = {}
  name = 0x0000000103a0d6d0
  super = 0x0000000103a4faf0
  parameters = 0x0000000103a08010
  types = 0x0000000103a562d0
  instance = 0x0000000000000000
  size = 16
  abstract = '\0'
  mutabl = '\0'
  pointerfree = '\x01'
  nfields = 1
  ninitialized = 1
  alignment = 8
  uid = 144
  struct_decl = 0x00000003082953f0
  ditype = 0x0000000000000000
  fields = {}
}
(lldb) p ((Type*)((jl_datatype_t*)jl_typeof(args[1]))->struct_decl)->dump()
%ASCIIString = type { %jl_value_t* }
(lldb) p jl_(((jl_datatype_t*)jl_typeof(args[1]))->name)
TypeName(name=:ASCIIString, module=Core, names=svec(:data), primary=ASCIIString, cache=svec(), linearcache=svec(), uid=143)
(lldb) p jl_(((jl_datatype_t*)jl_typeof(args[1]))->parameters)
svec()
(lldb) p jl_(((jl_datatype_t*)jl_typeof(args[1]))->types)
svec(ByteVec)
```

it appears that the definition of ByteVec is corrupted and `pointerfree` is unexpectedly set to `false` at this point:
```
(gdb) p jl_(0x7ffdf2c572b0)
ByteVec
$19 = void
(gdb) p *((jl_datatype_t*)0x7ffdf2c572b0)
$22 = {fieldptr0 = 0x7ffdf2c572b0, name = 0x7ffdf2c5ce10, super = 0x7ffdf2c57310, parameters = 0x7ffdf2c58010, types = 0x7ffdf2c66130, 
  instance = 0x0, size = 16, abstract = 0 '\000', mutabl = 0 '\000', pointerfree = 0 '\000', nfields = 1, ninitialized = 1, 
  alignment = 8, uid = 103, struct_decl = 0xa81f600, ditype = 0x0, fields = 0x7ffdf2c57300}
(gdb) p ((jl_datatype_t*)0x7ffdf2c572b0)->fields[0]
$23 = {offset = 0, size = 16, isptr = 0}
```

File: src/julia.h (hunk 2 above)
Comment (vtjnash) [inline src/julia.h]: you can't just randomly change information that codegen has decided to cache here and expect the rest of the codegen system to hold together.

**Comment by vtjnash** [comment]
just a thought: now that tuples are inline immutable, it might be interesting to just declare this:
```
immutable ASCIIString <: AbstractString
  data::NTuple{UInt8}
end
```
and then make jeff go optimize it (for example #11187)

File: src/julia.h (hunk 2 above)
Comment (StefanKarpinski) [inline src/julia.h]: What code would be generated by the time

https://github.com/JuliaLang/julia/blob/ed760e40e972fc1c6bee81139c9dc919fc5a790f/src/init.c#L1300

happens? Still, it entirely true that just setting `pointerfree = false` might not work – but it did seem to work in a previous incarnation of this pull-request, without any comments to the contrary from anyone who looked at it. Which might simply have been due to no one looking at it.

File: src/alloc.c (hunk 12 above)
Comment (StefanKarpinski) [inline src/alloc.c]: It's not returning a `jl_value_t` – `jl_bytevec_struct_t` is a C struct not a Julia value. The `ByteVec` type happens to have as its data a value that this struct is well-suited to manipulating.

File: src/array.c (hunk 13 above)
Comment (StefanKarpinski) [inline src/array.c]: Thanks, I'll add that and see if it helps.

**Comment by StefanKarpinski** [comment]
> just a thought: now that tuples are inline immutable...

Yes, @JeffBezanson had the same thought. I was hoping to get somewhere with this approach before having to wait for that optimization, but maybe it's better just to do it that way. Another thought is to just have a ByteVec type that always has a buffer but avoids all the overhead of Arrays. But we've already got too many vector-like types around, so that's kind of unappealing.

File: src/julia.h (hunk 2 above)
Comment (vtjnash) [inline src/julia.h]: the layout of an ASCIIString is already computed

File: src/alloc.c (hunk 12 above)
Comment (vtjnash) [inline src/alloc.c]: it's weird because it's got a gc pointer it in, but it's invisible to the gc when written this way

File: src/julia.h (hunk 2 above)
Comment (StefanKarpinski) [inline src/julia.h]: Where?

File: src/alloc.c (hunk 12 above)
Comment (StefanKarpinski) [inline src/alloc.c]: I agree that it's a bit weird, but this was the easiest way to go. I'm open to alternative suggestions.

File: src/julia.h (hunk 2 above)
Comment (vtjnash) [inline src/julia.h]: https://github.com/JuliaLang/julia/pull/11235/files#diff-4cbc8eb916240b8a07540daa4316d1c5R252

File: src/alloc.c (hunk 12 above)
Comment (vtjnash) [inline src/alloc.c]: moving the code to that creates the jl_value_t tag would seem to make sense to me

**Comment by simonster** [comment]
This blog post may be of some interest: https://blog.mozilla.org/ejpbruel/2012/02/06/how-strings-are-implemented-in-spidermonkey-2/

**Comment by ScottPJones** [comment]
@simonster Great post... interesting that 1) they consider string handling performance so important that they are willing to have some extra complexity 2) \0 termination is only guaranteed for certain types of strings

**Comment by vtjnash** [comment]
My tl;dr interpretation is they merge our ByteString/ByteVec/RopeString/SubString types into a single unified interface called a JavaScript string, and use copy-on-read to ensure O(n) efficiency of the common += operations, while preserving memory efficiency and \0 termination at all user-visible points.
